File: pythonlibrary/aibrd/chunker.py
import os
from .models.brd import RawBRD, BRDChunk

MAX_TOKENS = int(os.getenv("AIBRD_CHUNK_TOKENS", "6000"))
CHARS_PER_TOKEN = 4


def chunk_brd(brd: RawBRD, max_tokens: int = MAX_TOKENS) -> list[BRDChunk]:
    max_chars = max_tokens * CHARS_PER_TOKEN

    if len(brd.text) <= max_chars:
        return [BRDChunk(text=brd.text, index=0, total=1)]

    paragraphs = brd.text.split("\n\n")
    chunks: list[str] = []
    current = ""
    overlap = ""

    for para in paragraphs:
        candidate = f"{current}\n\n{para}" if current else para
        if len(candidate) > max_chars and current:
            chunks.append(overlap + current)
            overlap = current.split("\n\n")[-1] + "\n\n"
            current = para
        else:
            current = candidate

    if current:
        chunks.append(overlap + current)

    return [
        BRDChunk(text=text, index=i, total=len(chunks))
        for i, text in enumerate(chunks)
    ]

File: src/core/chunker.ts
import * as vscode from 'vscode'
import { BRDChunk, RawBRD } from './models/brd'

function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4)
}

export function chunkBRD(brd: RawBRD): BRDChunk[] {
  const config = vscode.workspace.getConfiguration('aibrd')
  const maxTokens: number = config.get('maxChunkTokens') ?? 6000
  const maxChars = maxTokens * 4

  if (brd.text.length <= maxChars) {
    return [{ text: brd.text, index: 0, total: 1 }]
  }

  const paragraphs = brd.text.split(/\n{2,}/)
  const chunks: BRDChunk[] = []
  let current = ''
  let overlap = ''

  for (const para of paragraphs) {
    const candidate = current ? `${current}\n\n${para}` : para
    if (estimateTokens(candidate) > maxTokens && current) {
      chunks.push({ text: overlap + current, index: chunks.length, total: 0 })
      // carry last paragraph as overlap for context continuity
      overlap = current.split('\n\n').slice(-1)[0] + '\n\n'
      current = para
    } else {
      current = candidate
    }
  }

  if (current) {
    chunks.push({ text: overlap + current, index: chunks.length, total: 0 })
  }

  return chunks.map(c => ({ ...c, total: chunks.length }))
}
